### PR TITLE
add postgresql functions concept

### LINKED ISSUE #4589: [Concept Entry] Postgresql functions
### Reason/inspiration (optional)

We would like a new entry on the `Functions` concept under PostgreSQL. The entry should go in a new file under `docs/content/postgresql/concepts/functions/functions.md`.

The entry should include:

- An introductory paragraph that describes the concept
- A �Syntax� section that includes the syntax for the concept (if applicable)
- An �Example� section that shows the concept in use

Please refer to the [concept entry template](https://github.com/Codecademy/docs/blob/main/documentation/concept-entry-template.md), [content standards](https://github.com/Codecademy/docs/blob/main/documentation/content-standards.md), and [markdown style guide](https://curriculum-documentation.codecademy.com/content-guidelines/markdown-style-guide/) when working on the PR for this issue.

### Entry Type (select all that apply)

New Concept Entrty for Existing Topic

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [X] Labels added
- [ ] Issue is assigned

### PR DESCRIPTION
**Pull request by hakdo99**
### Description

Add a concept entry about PostgreSQL functions. The new file is functions.md.

### Issue Solved

 Closes #4589

### Type of Change

- Adding a new entry

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] I have linked any issues that are relevant to this PR in the `Issues Solved` section.

### CODE REVIEW

File: content/postgresql/concepts/functions/functions.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+Title: 'Functions'
+Description: 'A PostgreSQL function is a reusable SQL code block that performs operations and returns results.'
+Subjects:
+  - 'Data Science'
+  - 'Computer Science'
+Tags:
+  - 'PostgreSQL'
+  - 'Functions'
+CatalogContent:
+  - 'learn-postgresql'
```
Comment (SaviDahegaonkar) [inline content/postgresql/concepts/functions/functions.md:11]:
```suggestion
  - 'learn-postgresql'
```
Include only those course paths that are already available in the [catalog-content.md file](documentation/catalog-content.md). You can add one if you want it to get included in the PR.  Here `learn-postgresql` is not available in the catalog-content file so it makes it invalid to get included in PR.

File: content/postgresql/concepts/functions/functions.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  - 'paths/analyze-data-with-sql'
+---
+
+**`Functions`** in PostgreSQL, also known as Stored Procedures, are sets of SQL and procedural commands stored on the database server. They enable us to perform executions within the database within a single function that usually require several commands and round trips.
```
Comment (SaviDahegaonkar) [inline content/postgresql/concepts/functions/functions.md:15]:
```suggestion
**`Functions`** in PostgreSQL, also known as Stored Procedures, are sets of SQL and procedural commands stored on the database server. They enable us to perform executions within the database within a single function that usually requires several commands and round trips.
```

File: content/postgresql/concepts/functions/functions.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+RETURNS return_type AS $$
+   DECLARE
+      < variable_declaration >
+      [...]
+   BEGIN
+      < function_body >
+      [...]
+      RETURN { variable_name | value }
+   END; 
+   $$ LANGUAGE lang_name;
```
Comment (SaviDahegaonkar) [inline content/postgresql/concepts/functions/functions.md:34]:
```suggestion
CREATE [OR REPLACE] FUNCTION function_name (parameters) 
RETURNS return_type AS $$
   DECLARE
      < variable_declaration >
      [...]
   BEGIN
      < function_body >
      [...]
      RETURN { variable_name | value } ;
   END; 
   $$ LANGUAGE lang_name;
```
Added semicolon at line number 32 for correct syntax.

**Review by SaviDahegaonkar — CHANGES_REQUESTED** [review]
Hey @hakdo99 ,
The entry is nicely written, and I have suggested a few changes. Please make them asap.

Thanks,
Savi

**Comment by SaviDahegaonkar** [comment]
Hey @NelsonSanti ,
Can I know if you have implemented the suggested changes so I can make it available for a second review?

**Comment by NelsonSanti** [comment]
Hello @SaviDahegaonkar 
it seems you were mistaken,
This job is not assigned to me.
No worries and thank you!

**Comment by SaviDahegaonkar** [comment]
> Hello @SaviDahegaonkar it seems you were mistaken, This job is not assigned to me. No worries and thank you!

Yeah, I am really sorry for the misunderstanding.

**Comment by SaviDahegaonkar** [comment]
Hey @hakdo99 ,
Can I know if you have implemented the suggested changes so I can make it available for a second review?

Thanks,
Savi

File: content/postgresql/concepts/functions/functions.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+* In `parameters`, list out the parameters that the function takes in. A function can have zero or more parameters.
+
+* Define the datatype of the returned value after `returns`. The `return_type` could be a base, composite, domain type, or even reference a table column's type.
+
+* The `AS` keyword is used when defining a standalone function.
+
+* `function_body` contains the logic of the function that will be executed.
+
+* `language lang_name` defines the name of the language that the function is implemented in. Note that PostgreSQL supports many languages including  SQL, C, PL/pgSQL, or the name of a user-defined procedural language.
```
Comment (SaviDahegaonkar) [inline content/postgresql/concepts/functions/functions.md:50]:
```suggestion
- `function_name` refers to the name of the function.

- [OR REPLACE] option allows you to update an existing function.

- In `parameters`, list out the parameters that the function takes in. A function can have zero or more parameters.

- Define the datatype of the returned value after `returns`. The `return_type` could be a base, composite, domain type, or even reference a table column's type.

- The `AS` keyword is used when defining a standalone function.

- `function_body` contains the logic of the function that will be executed.

- `language lang_name` defines the name of the language that the function is implemented in. Note that PostgreSQL supports many languages including  SQL, C, PL/pgSQL, or the name of a user-defined procedural language.
```

File: content/postgresql/concepts/functions/functions.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    SELECT AVG(salary)
+    INTO average_salary
+    FROM employees;
+
+    RETURN average_salary;
+END;
+$$ LANGUAGE plpgsql;
+```
+
+Breaking down the syntax:
```
Comment (SaviDahegaonkar) [inline content/postgresql/concepts/functions/functions.md:81]:
```suggestion
Breaking down the code:
```

**Review by SaviDahegaonkar — APPROVED** [review]
Hey @hakdo99 ,
LGTM! Thank You for Contributing to docs :)

Thanks,
Savi

File: content/postgresql/concepts/functions/functions.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
+---
+Title: 'Functions'
+Description: 'A PostgreSQL function is a reusable SQL code block that performs operations and returns results.'
+Subjects:
+  - 'Data Science'
+  - 'Computer Science'
```
Comment (Radhika-okhade) [inline content/postgresql/concepts/functions/functions.md:6]:
```suggestion
  - 'Computer Science'
  - 'Data Science'
```

File: content/postgresql/concepts/functions/functions.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+Tags:
+  - 'PostgreSQL'
+  - 'Functions'
```
Comment (Radhika-okhade) [inline content/postgresql/concepts/functions/functions.md:9]:
```suggestion
  - 'Functions'
  - 'PostgreSQL'
```

File: content/postgresql/concepts/functions/functions.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+  - 'paths/analyze-data-with-sql'
+---
+
+**`Functions`** in PostgreSQL, also known as Stored Procedures, are sets of SQL and procedural commands stored on the database server. They enable us to perform executions within the database within a single function that usually requires several commands and round trips.
+
+PostgreSQL supports functions in various languages, such as SQL, PL/pgSQL, C, Python, and more.
+
+## Syntax
+
+Here's the basic syntax used to define a function:
```
Comment (Radhika-okhade) [inline content/postgresql/concepts/functions/functions.md:20]:
```suggestion
Here is the syntax used to define a function:
```

File: content/postgresql/concepts/functions/functions.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+$$ LANGUAGE plpgsql;
+```
+
+
+- The function name `avg_salary()` is declared following the command `CREATE FUNCTION`. It accepts zero parameters.
+- The data type of the returned value is `NUMERIC`
+- The code block containing the declaration and the logic of the function is placed between a pair of `$$`
+- In the declaration section, declare a variable `average_salary` that stores the number of films from the `employees` table.
+- In the body, use the `select into` statement to select the average salary of the employees and assign it to the `average_salary` variable. At the end, use the return statement to `average_salary`
+
```
Comment (SaviDahegaonkar) [inline content/postgresql/concepts/functions/functions.md:87]:
```suggestion
Here the `average_salary` variable is used to store the result of the AVG(salary) calculation, and the function will return the average salary.

```

File: content/postgresql/concepts/functions/functions.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+Call the function to get the average salary:
```
Comment (SaviDahegaonkar) [inline content/postgresql/concepts/functions/functions.md:88]:
```suggestion
To call the avg_salary() function and get the average salary, we can use the `SELECT` statement as follows:
```

**Review by SaviDahegaonkar — APPROVED** [review]

File: content/postgresql/concepts/functions/functions.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+CREATE [OR REPLACE] FUNCTION function_name(parameters) 
+RETURNS return_type AS $$
+DECLARE
+    <variable_name> <data_type>;
+BEGIN
+   <function_logic>;
+   RETURN <value_or_variable>;
+END;
+$$ LANGUAGE lang_name;
+
```
Comment (SaviDahegaonkar) [inline content/postgresql/concepts/functions/functions.md:32]:
```suggestion
```
```

**Review by SaviDahegaonkar — APPROVED** [review]

File: content/postgresql/concepts/functions/functions.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
```
Comment (SaviDahegaonkar) [inline content/postgresql/concepts/functions/functions.md:32]:
```suggestion
```pseudo
CREATE [OR REPLACE] FUNCTION function_name(parameters) 
RETURNS return_type AS $$
DECLARE
    <variable_name> <data_type>;
BEGIN
   <function_logic>;
   RETURN <value_or_variable>;
END;
$$ LANGUAGE lang_name;
```
```

**Review by SaviDahegaonkar — APPROVED** [review]

File: content/postgresql/concepts/functions/functions.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+- `[OR REPLACE]`: This option allows you to update an existing function, with the same name. If the function doesn't exist, PostgreSQL will create a new one.
+- `parameters`: These are the input parameters that the function accepts. 
+- `RETURNS return_type`: Specifies the type of value that the function will return.
+- `AS`: It is used to introduce the function body. 
+- `function_body`: It contains the logic of the function that will be executed when the function is called.
+- `LANGUAGE lang_name` : Defines the name of the language in which the function is written.
+
+## Example
+
+Suppose we have a table `employees` containing the employees' names and salaries, with the following records:
```
Comment (SaviDahegaonkar) [inline content/postgresql/concepts/functions/functions.md:43]:
```suggestion
Suppose we have a table `employees` containing employees names and salaries, with the following records:
```

**Review by SaviDahegaonkar — APPROVED** [review]

File: content/postgresql/concepts/functions/functions.md [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+$$ LANGUAGE lang_name;
+```
+
+- `function_name`: Refers to the name of the function.
+- `[OR REPLACE]`: This option allows you to update an existing function, with the same name. If the function doesn't exist, PostgreSQL will create a new one.
+- `parameters`: These are the input parameters that the function accepts. 
+- `RETURNS return_type`: Specifies the type of value that the function will return.
+- `AS`: It is used to introduce the function body. 
+- `function_body`: It contains the logic of the function that will be executed when the function is called.
+- `LANGUAGE lang_name` : Defines the name of the language in which the function is written.
```
[the remark this answers is not in the document]
Comment (codecademy-docs) [reply content/postgresql/concepts/functions/functions.md:40]:
```suggestion
- `function_name`: Refers to the name of the function.
- `[OR REPLACE]`: It updates an existing function with the same name. If the function does not exist, PostgreSQL will create a new one. 
- `parameters`: These are the input parameters that the function accepts. 
- `RETURNS return_type`: Specifies the type of value that the function will return.
- `AS`: It is used to introduce the function body. 
- `function_body`: This field contains the function's logic that will be executed when the function is called.
- `LANGUAGE lang_name`: Defines the name of the language in which the function is written.
```

[the remark this answers is not in the document]
Comment (Radhika-okhade) [reply content/postgresql/concepts/functions/functions.md:40]:
```suggestion
- `function_name`: Refers to the name of the function.
- `[OR REPLACE]`: It allows you to update an existing function with the same name. If the function does not exist, PostgreSQL will create a new one.
- `parameters`: These are the input parameters that the function accepts. 
- `RETURNS return_type`: Specifies the type of value that the function will return.
- `AS`: It is used to introduce the function body. 
- `function_body`: This field contains the function's logic that will be executed when called.
- `LANGUAGE lang_name`: Defines the name of the language in which the function is written.
```

File: content/postgresql/concepts/functions/functions.md [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+  - 'Computer Science'
+  - 'Data Science'
+Tags:
+  - 'Functions'
+  - 'PostgreSQL'
+CatalogContent:
+  - 'paths/analyze-data-with-sql'
+---
+
+**`Functions`** in PostgreSQL are reusable sets of SQL and procedural commands stored on the database server. They are also referred to as Stored Procedures. They enable us to perform complex operations within the database within a single function that usually requires several commands and round trips.
```
Comment (Radhika-okhade) [inline content/postgresql/concepts/functions/functions.md:14]:
```suggestion
**`Functions`** in PostgreSQL are reusable sets of SQL and procedural commands stored on the database server. They are also referred to as *Stored Procedures*. They enable us to perform complex operations within the database within a single function that usually requires several commands and round trips.
```

File: content/postgresql/concepts/functions/functions.md [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+$$ LANGUAGE lang_name;
+```
+
+- `function_name`: Refers to the name of the function.
+- `[OR REPLACE]`: It allows you to update an existing function with the same name. If the function does not exist, PostgreSQL will create a new one.
+- `parameters`: These are the input parameters that the function accepts. 
+- `RETURNS return_type`: Specifies the type of value that the function will return.
+- `AS`: It is used to introduce the function body. 
+- `function_body`: This field contains the function's logic that will be executed when called.
+- `LANGUAGE lang_name`: Defines the name of the language in which the function is written.
```
Comment (Radhika-okhade) [inline content/postgresql/concepts/functions/functions.md:40]:
```suggestion
- `function_name`: Specifies the function name.
- `[OR REPLACE]`: Allows updating an existing function with the same name. PostgreSQL will create a new function if the function does not exist
- `parameters`: These are the input parameters that the function accepts. 
- `RETURNS return_type`: Specifies the type of value that the function will return.
- `AS`: Introduces the function body. 
- `function_body`: This field contains the function's logic that will be executed when the function is called.
- `LANGUAGE lang_name`: Defines the name of the language in which the function is written.
```

File: content/postgresql/concepts/functions/functions.md [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+## Example
+
+Suppose we have a table `employees` containing employees names and salaries, with the following records:
+
+| employee | salary |
+|----------|--------|
+| Sarah    | 50000  |
+| Bailey   | 60000  |
+| Valeria  | 70000  |
+We would like to create a function that calculates the average salary of all employees in the table. Our function `avg_salary()` , can be defined as follows:
```
Comment (Radhika-okhade) [inline content/postgresql/concepts/functions/functions.md:51]:
```suggestion
We want to create a function that calculates the average salary of all employees in the table. The function `avg_salary()` can be defined as follows:
```

File: content/postgresql/concepts/functions/functions.md [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+    INTO average_salary
+    FROM employees;
+
+    RETURN average_salary;
+END;
+$$ LANGUAGE plpgsql;
+```
+Here the `average_salary` variable is used to store the result of the AVG(salary) calculation, where the function will return the average salary.
+
+To call the avg_salary() function and get the average salary, we can use the `SELECT` statement as follows:
```
Comment (Radhika-okhade) [inline content/postgresql/concepts/functions/functions.md:69]:
```suggestion
To call the `avg_salary()` function and get the average salary, use the `SELECT` statement as follows:
```

File: content/postgresql/concepts/functions/functions.md [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
+```sql
+SELECT avg_salary();
+```
+
+Thie above code will call the `avg_salary( )` function and return the calculated average salary of all the emplyees in the `employee` table.
```
Comment (Radhika-okhade) [inline content/postgresql/concepts/functions/functions.md:75]:
```suggestion
The above code will call the `avg_salary( )` function and return all the employees calculated average salaries in the `employee` table.
```

File: content/postgresql/concepts/functions/functions.md [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+    average_salary NUMERIC;
+BEGIN
+    SELECT AVG(salary)
+    INTO average_salary
+    FROM employees;
+
+    RETURN average_salary;
+END;
+$$ LANGUAGE plpgsql;
+```
```
Comment (Radhika-okhade) [inline content/postgresql/concepts/functions/functions.md:66]:
```suggestion
```

```

File: content/postgresql/concepts/functions/functions.md [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+## Example
+
+Suppose we have a table `employees` containing employees names and salaries, with the following records:
+
+| employee | salary |
+|----------|--------|
+| Sarah    | 50000  |
+| Bailey   | 60000  |
+| Valeria  | 70000  |
+We want to create a function that calculates the average salary of all employees in the table. The function `avg_salary()` can be defined as follows:
```
Comment (Radhika-okhade) [inline content/postgresql/concepts/functions/functions.md:51]:
```suggestion

We want to create a function that calculates the average salary of all employees in the table. The function `avg_salary()` can be defined as follows:
```

**Review by Radhika-okhade — APPROVED** [review]

**Pull request merged**